pkg/scale/uint128.go

```diff
@@ -0,0 +1,148 @@
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+package scale
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/big"
+)
+
+// Uint128 represents an unsigned 128 bit integer
+type Uint128 struct {
+	Upper uint64
+	Lower uint64
+}
+
+// MaxUint128 is the maximum uint128 value
+var MaxUint128 = &Uint128{
+	Upper: ^uint64(0),
+	Lower: ^uint64(0),
+}
+
+// MustNewUint128 will panic if NewUint128 returns an error
+func MustNewUint128(in interface{}, order ...binary.ByteOrder) (u *Uint128) {
+	u, err := NewUint128(in, order...)
+	if err != nil {
+		panic(err)
+	}
+	return
+}
+
+func padBytes(b []byte, order binary.ByteOrder) []byte {
+	for len(b) != 16 {
+		switch order {
+		case binary.BigEndian:
+			b = append([]byte{0}, b...)
+		case binary.LittleEndian:
+			b = append(b, 0)
+		}
+	}
+	return b
+}
+
+// NewUint128 is constructor for Uint128 that accepts an option binary.ByteOrder
+// option is only used when inputted interface{} is of type []byte
+// by default binary.LittleEndian is used for []byte since this is SCALE
+func NewUint128(in interface{}, order ...binary.ByteOrder) (u *Uint128, err error) {
+	switch in := in.(type) {
+	case *big.Int:
+		bytes := in.Bytes()
+		if len(bytes) < 16 {
+			bytes = padBytes(bytes, binary.BigEndian)
+		}
+		u = &Uint128{
+			Upper: binary.BigEndian.Uint64(bytes[:8]),
+			Lower: binary.BigEndian.Uint64(bytes[8:]),
+		}
+	case []byte:
+		var o binary.ByteOrder = binary.LittleEndian
+		if len(order) > 0 {
+			o = order[0]
+		}
+		if len(in) < 16 {
+			in = padBytes(in, o)
+		}
+		u = &Uint128{
+			Upper: o.Uint64(in[8:]),
+			Lower: o.Uint64(in[:8]),
+		}
+	default:
+		err = fmt.Errorf("unsupported type: %T", in)
+	}
+	return
+}
+
+// Bytes returns the Uint128 in little endian format by default.  A variadic paramter
+// order can be used to specify the binary.ByteOrder used
+func (u *Uint128) Bytes(order ...binary.ByteOrder) (b []byte) {
+	var o binary.ByteOrder = binary.LittleEndian
+	if len(order) > 0 {
+		o = order[0]
+	}
+	b = make([]byte, 16)
+	switch o {
+	case binary.LittleEndian:
+		o.PutUint64(b[:8], u.Lower)
+		o.PutUint64(b[8:], u.Upper)
+		b = u.trimBytes(b, o)
+	case binary.BigEndian:
+		o.PutUint64(b[:8], u.Upper)
+		o.PutUint64(b[8:], u.Lower)
+		b = u.trimBytes(b, o)
+	}
+	return
+}
+
+// Cmp returns 1 if the receiver is greater than other, 0 if they are equal, and -1 otherwise.
+func (u *Uint128) Compare(other *Uint128) int {
+	switch {
+	case u.Upper > other.Upper, u.Lower > other.Lower:
+		return 1
+	case u.Upper < other.Upper, u.Lower < other.Lower:
+		return -1
+	default:
+		return 0
+	}
```

Comment:
are the cases ands or ors? I don't think this is correct, what if `u.Upper == other.Upper` but `u.Upper > other.Upper`? then `u` is greater than `other` but I don't think this checks that case?

Reply:
OR cases.  I'll double check the original implementation.

Reply:
```
func (u *Uint128) Cmp(other *Uint128) int {
	if u.Upper > other.Upper {
		return 1
	}

	if u.Upper < other.Upper {
		return -1
	}

	if u.Lower > other.Lower {
		return 1
	}

	if u.Lower < other.Lower {
		return -1
	}

	return 0
}
```

Reply:
I see what you mean here.  I'll rewrite.